### ISSUE TITLE
how to use cargo in the benchmarks game

### ISSUE THREAD

**Issue by BurntSushi** [TeXitoi/benchmarksgame-rs#3]
I just went through the process of getting the regex benchmark updated (because it recently got a lot of optimizations), and this required Cargo because `regex` now has four total dependencies: `regex-syntax`, `aho-corasick`, `memchr` and `libc`.

It was a little tricky to get working because the maintainer still needs to be able to run `rustc` on the submitted benchmark.

Here is the thread: https://alioth.debian.org/tracker/index.php?func=detail&aid=315089&group_id=100815&atid=413122

The approach we finally settled on was to set up a blank Cargo project with `regex` as a dependency, `cargo build --release` it, and then use `rustc` with `-L` with the `./target/release/deps` directory:

```
$ ls
regex-dna.rs
$ cargo new regex-deps
$ printf '[dependencies]\nregex = "=0.1.38"\n' >> regex-deps/Cargo.toml 
$ cargo build --release --manifest-path regex-deps/Cargo.toml 
$ rustc -C opt-level=3 -C lto -L ./regex-deps/target/release/deps/ regex-dna.rs -o regex-dna
$ ./regex-dna < /data/regex-dna.fasta
```

This does fail on 32 bit unfortunately, as can be seen here at the bottom of the list: http://benchmarksgame.alioth.debian.org/u32/performance.php?test=regexdna I filed a bug with rustc: https://github.com/rust-lang/rust/issues/26541

I'm hoping this info is useful enough so that we can:
1. Get this build process setup in this repo so that the `regex` benchmark can be maintained here.
2. Possibly do something similar for other benchmarks that are allowed to use a library.

**Comment by BurntSushi** [comment]
I talked to @alexcrichton on IRC, and he seems to think the 32 bit problem is an LTO related bug. Perhaps we could open a new thread in the benchmark game forum and ask the maintainer to not use LTO? (I guess this might be a question that must be answered for all Rust benchmarks.)

**Comment by BurntSushi** [comment]
@TeXitoi The maintainer has stated that we can selectively disable LTO on certain benchmarks. I'm in favor of disabling it for `regex-dna` at least, but we can't go back and forth on it with the Benchmark Game as the intermediate. Is there anyone else I should speak with?

**Comment by TeXitoi** [comment]
I'm OK to disable lto everywhere. 

I'll look for the cargo installation with dependencies.

Sorry I'm quite busy now, I don't know when I'll have time to do that.

**Comment by llogiq** [comment]
I think I have it working now. I'll push a PR shortly.

**Comment by TeXitoi** [comment]
closed by #13
